### PR TITLE
Add kernel modules documentation

### PR DESCRIPTION
**Pull request by puneetse**
This PR adds a new document describing kernel modules in CL and loading out-of-tree modules, which has been a popular question on mailing lists and IRC.

### CODE REVIEW

**Comment by intelkevinputnam** [comment]
@puneetse Can you add kernel-modules to the TOC tree in maintenance.rst and update your PR?

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+##############
+Kernel modules are additional pieces of software capable of interacting with 
+the kernel to add functionality, such as a hardware driver. Kernel modules 
+may already be part of the Linux source tree (in-tree) or may come from an 
+external source, such as a directly from vendor (out-of-tree).  
+
+In many cases kernel modules are already available through |CL|. 
+In other cases, a kernel module may not exist but can be requested to be 
+enabled in the |CL| kernel. Finally, a situation may call for manually 
+building and loading out-of-tree kernels.
```
Comment (bwarden) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: ...out-of-tree modules.

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 2]
```diff
@@ -0,0 +1,186 @@
+.. _kernel-modules:
+
+Kernel Modules 
+##############
+Kernel modules are additional pieces of software capable of interacting with
```
Comment (bwarden) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: ...capable of being inserted into the kernel...

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+Check if the module is already available through Clear Linux
+============================================================
+|CL| comes with many upstream kernel modules available for use.  If 
+you require a kernel module, be sure to check whether it is already available in |CL| first. 
+
+
+You can search for kernel modules using the :command:`swupd search` command. 
+See :ref:`swupd-search` for more information. 
+
```
Comment (bwarden) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: Maybe an example here, emphasis on the .ko extension to constrain the search.

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+   parameters and reboot the system. 
+
+   All kernel modules from |CL| have been signed to enforce kernel security. 
+   However, out-of-tree modules break this chain of trust so this mechanism 
+   needs to be disabled.
+  
+    .. code-block:: bash
+
+        sudo mkdir -p /etc/kernel/cmdline.d
+        sudo echo "module.sig_unenforce" > /etc/kernel/cmdline.d/load-modules.conf
```
Comment (bwarden) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: sudo is useless here, as it applies only to the echo and not the redirect. Could rewrite as:
echo "module.sig_unenforce" | sudo tee /etc/kernel/cmdline.d/load-modules.conf

Comment (bwarden) [reply source/clear-linux/guides/maintenance/kernel-modules.rst]: Might also rename the example conf file to "allow-unsigned-modules.conf" to be more explicit

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 5]
```diff
@@ -0,0 +1,192 @@
+.. _kernel-modules:
+
+Kernel Modules 
+##############
+Kernel modules are additional pieces of software capable of being inserted 
+into the Linux kernel to add functionality, such as a hardware driver. 
+Kernel modules may already be part of the Linux source tree (in-tree) or may 
+come from an external source, such as a directly from vendor (out-of-tree).  
+
+In many cases kernel modules are already available through |CL|.
```
Comment (intelkevinputnam) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: line 10: change |CL| to |CL-ATTR| (first reference).

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 6]
```diff
@@ -0,0 +1,192 @@
+.. _kernel-modules:
```
Comment (intelkevinputnam) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: Title needs to be more focused. Suggest: Loading an out-of-tree Kernel Module.

Comment (intelkevinputnam) [reply source/clear-linux/guides/maintenance/kernel-modules.rst]: As a result, I recommend a new intro with th explanation of kernel modules along side a description of what the article is going to show you how to do.

Comment (MCamp859) [reply source/clear-linux/guides/maintenance/kernel-modules.rst]: > Title needs to be more focused. Suggest: Loading an out-of-tree Kernel Module.

Please change title to "Load an out-of-tree kernel module" for consistency with other help topics (no -ing words and sentence case).

Comment (puneetse) [reply source/clear-linux/guides/maintenance/kernel-modules.rst]: > Title needs to be more focused. Suggest: Loading an out-of-tree Kernel Module.

I'm not sure I agree for the top level title. The document is broader about kernel modules in CL. Loading out of tree modules is just one solution in the document.

**Comment by intelkevinputnam** [comment]
@MCamp859 can you take another look?

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+   out-of-tree modules.
+
+   .. code-block:: bash
+
+      $ uname -r
+      4.XX.YY-ZZZZ.native
+      # Ensure *.native* is in the kernel name
+
+#. Install the `linux-dev` bundle to obtain the kernel headers, which are
+   required for compiling kernel modules
```
Comment (MCamp859) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: Add . at end of sentence.

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+   .. code-block:: bash
+
+      sudo swupd bundle-add linux-dev
+
+#. Follow instructions from the kernel module source code to compile the 
+   kernel module
```
Comment (MCamp859) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: Add . at end of sentence.

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+.. code-block:: bash
+
+   man modprobe.d
+
+Optional: Configure kernel modules to load at boot
+--------------------------------------------------
+
+The :file:`/etc/modules-load.d` configuration directory can be used to 
+specify kernel modules that should be automatically loaded at boot.
```
Comment (MCamp859) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: Suggest rewriting in active voice like the first Optional item. Something like: 
Create a :file:`/etc/modules-load.d` configuration directory to specify kernel modules to automatically load at boot.

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+modules having dependencies, a method of specifying what options are 
+to be used with individual modules is useful. This can be done with 
+configuration files under the :file:`/etc/modprobe.d` directory. 
+
+.. code-block:: bash
+
+   sudo mkdir /etc/modprobe.d
+
+All files underneath the :file:`/etc/modprobe.d` directory 
+which end with the :file:`.conf` extension specify module options to use when
```
Comment (MCamp859) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: change "which" to "that"

File: source/clear-linux/guides/maintenance/kernel-modules.rst [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+The :file:`/etc/modules-load.d` configuration directory can be used to 
+specify kernel modules that should be automatically loaded at boot.
+
+.. code-block:: bash
+
+   sudo mkdir /etc/modules-load.d
+
+All files underneath the :file:`/etc/modules-load.d` directory 
+which end with the :file:`.conf` extension contain a list of module names
```
Comment (MCamp859) [inline source/clear-linux/guides/maintenance/kernel-modules.rst]: change "which" to "that"

**Comment by intelkevinputnam** [comment]
Thanks @MCamp859! Much appreciated!

**Pull request merged**